File: 653.TwoSumIV-InputisaBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function(root, k) { // Considering BST is a sorted array when tranversing with DFS (inOrderTranversing), so convert it to an sorted array then solve the 2Sum problem
    function convertBSTtoArray(node) {
        const stack = [];
        const list = [];
        let current = node;
        let cursor = 0;
        while(true) {
            if (current) {
                stack.push(current);
                current = current.left;
            } else if (stack.length) {
                const node = stack.pop();
                list.push(node.val);
                current = node.right;
            } else {
                break;
            }
        }
        return list;
    }
    const list = convertBSTtoArray(root);
    let i = 0;
    let j = 1;
    const len = list.length;
    while (i < j && j < len) {
        if (list[i] + list[j] === k) {
            return true;   
        }
        j++;
        if (j === len) {
            i++;
            j = i + 1;
        }
    }
    return false;
};

File: Golang/Tree/653.Two-sum-iv-input-is-a-bst.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findTarget(root *TreeNode, k int) bool {
	hashmap := map[int]bool{}
	current := root
	stack := []*TreeNode{}
	for {
		if current != nil {
			stack = append(stack, current)
			current = current.Left
		} else if len(stack) > 0 {
			node := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if hashmap[k-node.Val] {
				return true
			}
			hashmap[node.Val] = true
			current = node.Right
		} else {
			break
		}
	}
	return false
}
